restapi/src/main/java/io/stargate/web/docsapi/service/query/search/resolver/BaseResolver.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.stargate.web.docsapi.service.query.search.resolver;
+
+import com.bpodgursky.jbool_expressions.Expression;
+import com.bpodgursky.jbool_expressions.Literal;
+import com.bpodgursky.jbool_expressions.rules.RuleSet;
+import io.stargate.web.docsapi.service.ExecutionContext;
+import io.stargate.web.docsapi.service.query.FilterExpression;
+
+/**
+ * Base resolver knows what {@link DocumentsResolver} should be created for the given {@link
+ * Expression}.
+ */
+public final class BaseResolver {
+
+  private BaseResolver() {}
+
+  /**
+   * Resolves the document resolver without any parent.
+   *
+   * @param expression {@link Expression}
+   * @return DocumentsResolver
+   */
+  public static DocumentsResolver resolve(
+      Expression<FilterExpression> expression, ExecutionContext context) {
+    return resolve(expression, context, null);
+  }
+
+  /**
+   * Resolves the document resolver with optional parent.
+   *
+   * @param expression {@link Expression}
+   * @param parent parent or <code>null</code>
+   * @return DocumentsResolver
+   */
+  public static DocumentsResolver resolve(
+      Expression<FilterExpression> expression, ExecutionContext context, DocumentsResolver parent) {
+
+    // if we are hitting the literal TRUE, then return parent
+    if (Literal.EXPR_TYPE.equals(expression.getExprType())) {
+      return parent;
+    }
+
+    // if we have only one Filter expression, then to SingleExpression resolver
+    if (FilterExpression.EXPR_TYPE.equals(expression.getExprType())) {
+      return SingleExpressionResolver.resolve((FilterExpression) expression, context, parent);
+    }
+
+    // for now in every other case go for the CNF resolver
+    Expression<FilterExpression> cnfExpression = RuleSet.toCNF(expression);
```

Comment:
This can reorder filters even when it does not materially change the expression (i.e. even when the expression is already in CNF form)

Reply:
For now, since we only allow `AND` at request level, we could just skip the CNF conversion, I guess, and deal with it later, when we add `ORs`.

Reply:
Not actually converting to CNF for now might also be preferable because the `CnfResolver` assumes (e.g. in its `getFilterExpressions(...)` method) that sub-expressions of `AND` are plain filter expressions, but in general they could be `ORs` of filter expressions (with negation). So, we do not really take full advantage of the CNF form now. Perhaps we could defer doing this conversion until we tackle `ORs`. WDYT?

Reply:
Of course, we if go without CNF conversion, it might be worth renaming `CnfResolver` to `AndFilterResolver` or something like that.

As for me, I'd just move all resolution code for all sub-cases into `BaseResolver` as I believe `SingleExpressionResolver` is redundant (see other comments about that).

Reply:
Two takeaways:
1. yes I can remove the SingleExpressionResolver, as actually I can wrap a single expression into and `And`!
2. I am keeping the CNF stuff, as this is basis for the ORs imo, although at the moment it's not really helping. However, if we did not do it like this from the start we would never hit this reordering and we would never think about this selectivity support. So let's keep it as it is and solve the problems.

Reply:
Done.